packages/@aws-cdk/pipelines/lib/private/application-security-check.ts

```diff
@@ -0,0 +1,112 @@
+   * }
+   */
+  public readonly preApproveLambda: lambda.Function;
+  /**
+   * A CodeBuild Project that runs a security diff on the application stage.
+   *
+   * - If the diff registers no security changes, CodeBuild will invoke the
+   *   pre-approval lambda and approve the ManualApprovalAction.
+   * - If changes are detected, CodeBuild will exit into a ManualApprovalAction
+   */
+  public readonly cdkDiffProject: codebuild.Project;
+
+  constructor(scope: Construct, id: string) {
+    super(scope, id);
+
+    this.preApproveLambda = new lambda.Function(this, 'CDKPipelinesAutoApprove', {
+      handler: 'index.handler',
+      runtime: lambda.Runtime.PYTHON_3_8,
+      code: lambda.Code.fromAsset(path.resolve(__dirname, '../lambda')),
+      timeout: Duration.seconds(30),
+    });
+
+    this.preApproveLambda.addToRolePolicy(new iam.PolicyStatement({
+      actions: ['codepipeline:GetPipelineState', 'codepipeline:PutApprovalResult'],
+      conditions: {
+        StringEquals: {
+          'aws:ResourceTag/SECURITY_CHECK': 'ALLOW_APPROVE',
+        },
+      },
+      resources: ['*'],
+    }));
+
+    const assemblyPath = 'assembly-$STACK_NAME-$STAGE_NAME/';
+    const invokeLambda =
+      'aws lambda invoke' +
+      ` --function-name ${this.preApproveLambda.functionName}` +
+      ' --invocation-type Event' +
+      ' --payload "$payload"' +
+      ' lambda.out';
+
+    this.cdkDiffProject = new codebuild.Project(this, 'CDKSecurityCheck', {
+      buildSpec: codebuild.BuildSpec.fromObject({
+        version: 0.2,
+        phases: {
+          build: {
+            commands: [
+              'npm install -g aws-cdk',
+              // $CODEBUILD_INITIATOR will always be Code Pipeline and in the form of:
+              // "codepipeline/example-pipeline-name-Xxx"
+              'payload="$(node -pe \'JSON.stringify({ "PipelineName": process.env.CODEBUILD_INITIATOR.split("/")[1], "StageName": process.env.STAGE_NAME, "ActionName": process.env.ACTION_NAME })\' )"',
+              // ARN: "arn:aws:codebuild:$region:$account_id:build/$project_name:$project_execution_id$"
+              'ARN=$CODEBUILD_BUILD_ARN',
+              'REGION="$(node -pe \'`${process.env.ARN}`.split(":")[3]\')"',
+              'ACCOUNT_ID="$(node -pe \'`${process.env.ARN}`.split(":")[4]\')"',
+              'PROJECT_NAME="$(node -pe \'`${process.env.ARN}`.split(":")[5].split("/")[1]\')"',
+              'PROJECT_ID="$(node -pe \'`${process.env.ARN}`.split(":")[6]\')"',
+              // Manual Approval adds 'http/https' to the resolved link
+              'export LINK="$REGION.console.aws.amazon.com/codesuite/codebuild/$ACCOUNT_ID/projects/$PROJECT_NAME/build/$PROJECT_NAME:$PROJECT_ID/?region=$REGION"',
+              // Run invoke only if cdk diff passes (returns exit code 0)
+              // 0 -> true, 1 -> false
+              `if cdk diff -a "${assemblyPath}" --security-only --fail; then ${invokeLambda}; export MESSAGE='No changes detected.'; ` +
```

Comment:
To make this easier to read (more like what you'd write in a Bash script where you didn't have to put everything on one line), you could write something readable in TypeScript and then transform that into the long, squished line that the CodeBuild spec requires. I typically do an array that I join:

```ts
  'export LINK=...',
  [
     'if cdk diff; then',
     ` ${invokeLambda};`
     ' export message;',
     'else',
     '  export message;'
     'fi'
   ].join(' '),
});
```

(Still not *entirely* what you'd write in Bash since you need to remember to put `;` in the right place, but at least more readable. We COULD make a function that would insert the `;` in the right place, everywhere at the end of a line except after `then` and `else`, depending on how fancy you want to get...)